File: mediapipe/tasks/python/components/processors/classifier_options.py
# Copyright 2022 The TensorFlow Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Classifier options data class."""

import dataclasses
from typing import Any, List, Optional


@dataclasses.dataclass
class ClassifierOptions:
  """Options for classification processor.

  Attributes:
    display_names_locale: The locale to use for display names specified through
      the TFLite Model Metadata.
    max_results: The maximum number of top-scored classification results to
      return.
    score_threshold: Overrides the ones provided in the model metadata. Results
      below this value are rejected.
    category_allowlist: Allowlist of category names. If non-empty, detection
      results whose category name is not in this set will be filtered out.
      Duplicate or unknown category names are ignored. Mutually exclusive with
      `category_denylist`.
    category_denylist: Denylist of category names. If non-empty, detection
      results whose category name is in this set will be filtered out. Duplicate
      or unknown category names are ignored. Mutually exclusive with
      `category_allowlist`.
  """

  display_names_locale: Optional[str] = None
  max_results: Optional[int] = None
  score_threshold: Optional[float] = None
  category_allowlist: Optional[List[str]] = None
  category_denylist: Optional[List[str]] = None

  def __eq__(self, other: Any) -> bool:
    """Checks if this object is equal to the given object.

    Args:
      other: The object to be compared with.

    Returns:
      True if the objects are equal.
    """
    if not isinstance(other, ClassifierOptions):
      return False
    return (
        self.display_names_locale == other.display_names_locale
        and self.max_results == other.max_results
        and self.score_threshold == other.score_threshold
        and self.category_allowlist == other.category_allowlist
        and self.category_denylist == other.category_denylist
    )

File: mediapipe/tasks/web/components/processors/classifier_options.ts
/**
 * Copyright 2022 The MediaPipe Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {ClassifierOptions as ClassifierOptionsProto} from '../../../../tasks/cc/components/processors/proto/classifier_options_pb';
import {ClassifierOptions} from '../../../../tasks/web/core/classifier_options';

/**
 * Converts a ClassifierOptions object to its Proto representation, optionally
 * based on existing definition.
 * @param options The options object to convert to a Proto. Only options that
 *     are expliclty provided are set.
 * @param baseOptions A base object that options can be merged into.
 */
export function convertClassifierOptionsToProto(
    options: ClassifierOptions,
    baseOptions?: ClassifierOptionsProto): ClassifierOptionsProto {
  const classifierOptions =
      baseOptions ? baseOptions.clone() : new ClassifierOptionsProto();
  if (options.displayNamesLocale !== undefined) {
    classifierOptions.setDisplayNamesLocale(options.displayNamesLocale);
  } else if (options.displayNamesLocale === undefined) {
    classifierOptions.clearDisplayNamesLocale();
  }

  if (options.maxResults !== undefined) {
    classifierOptions.setMaxResults(options.maxResults);
  } else if ('maxResults' in options) {  // Check for undefined
    classifierOptions.clearMaxResults();
  }

  if (options.scoreThreshold !== undefined) {
    classifierOptions.setScoreThreshold(options.scoreThreshold);
  } else if ('scoreThreshold' in options) {  // Check for undefined
    classifierOptions.clearScoreThreshold();
  }

  if (options.categoryAllowlist !== undefined) {
    classifierOptions.setCategoryAllowlistList(options.categoryAllowlist);
  } else if ('categoryAllowlist' in options) {  // Check for undefined
    classifierOptions.clearCategoryAllowlistList();
  }

  if (options.categoryDenylist !== undefined) {
    classifierOptions.setCategoryDenylistList(options.categoryDenylist);
  } else if ('categoryDenylist' in options) {  // Check for undefined
    classifierOptions.clearCategoryDenylistList();
  }
  return classifierOptions;
}
